lib/runnable.js

```diff
@@ -229,14 +208,14 @@ Runnable.prototype.clearTimeout = function() {
  */
 Runnable.prototype.resetTimeout = function() {
   var self = this;
-  var ms = this.timeout() || 1e9;
```

Comment:
What was the purpose of `1e9`?

Reply:
`"so someone could not create an async test with a timeout of 0 and have Node exit silently"` (details https://github.com/mochajs/mocha/commit/7c1139e8b47f817dd05508129d8134a9eafe027b).

Reply:
```js
describe('something', function () {
  it('should work', function (done) {
    this.timeout(0);
  })
});
```
Above test does not work, neither with current master nor with your branch. Node exactly does exit silently.
I don't know what's wrong, neither why you just haved removed `1e9`.

Reply:
@juergba 

> neither why you just haved removed 1e9.

So I understood the previous code to mean, if `enableTimeout(true)` and `timeout(0)` then use `1e9` default value for timeout. However now 0 == disabled so this is default value is no longer required (I thought).

> I don't know what's wrong

We have disabled the timeout errors (via early returns) for `timeout(0)` (see `resetTimeout`) so in this scenario would not get the timeout exceeded error (which would say `'ensure "done()" is called'`). 

So as I understand the issue is that if you disable timeouts but run a test which would never complete, we fail silently. 

Couple of attempts:
- If I removed all the early exits inside `resetTimeout` and put the `1e9` default back then your example test will fail correctly, but will take a hell of a long time to fail, so not a good idea.
- If I have the early return inside the `timer = setTimeout` return a timeout error, then it will throw all the time.

Im not quite sure how to resolve this. Any ideas are welcome. If we think its a bigger issue we could move into a new Issue as this one was never meant to re-define the timeout behaviour to this extent.

Reply:
IMO, the following test _should_ not exit silently:  

```js
describe('something', function () {
  it('should work', function (done) {
    this.timeout(0);
  })
});
```

But there's not much to be done here.  Waiting `1e9` ms is just kind of dumb.  A more elegant solution would be to leverage `async_hooks` or `AsyncLocalStorage` to _check_, when the test exits, if the user has put a task into the event loop (and `done()` hasn't yet been called).  If no, then we could throw an exception.  But we can't support that behavior in a browser without some kind of an adapter (e.g., Angular's Zones).

I guess I don't think you should do anything about it unless there are open issues about silent exits.

There's a contract here with `done`: either you call it right away (or throw an exception), or you put a task in the event loop which will either a) call it later, or b) throw an exception.  Maybe the correct solution for now is to just make sure it's in writing.

Reply:
to be clear, I think the removal of `1e9` is fair, because it wasn't working as intended _at least_ since the addition of `_enableTimeouts`.

Reply:
I found issue #2537 (from 2016) where Node aborts execution with exitCode==0.

> Waiting 1e9 ms is just kind of dumb.

Everything is better than abort with exitCode==0, and actually it's the behaviour I would expect.

IMO you should first fix/clarify #2537 (in a different PR) making sure we really don't need `_enableTimeouts` anymore. Then fix/merge this one.
I leave it upto you. I backout of this PR.

Reply:
A different approach is to go back to my original solution ([commit](https://github.com/mochajs/mocha/pull/4260/commits/d091cdc3f1a4b25783689302e198b2482cd8f30b)) which (1) fixed the intended bug, (2) did not remove `this.enableTimeout` api.